drivers/disk/flashdisk.c

```diff
@@ -279,16 +337,35 @@ static const struct disk_operations flash_disk_ops = {
 	.ioctl = disk_flash_access_ioctl,
 };
 
-static struct disk_info flash_disk = {
-	.name = CONFIG_DISK_FLASH_VOLUME_NAME,
-	.ops = &flash_disk_ops,
+#define DT_DRV_COMPAT zephyr_flash_disk
+
+#define PARTITION_PHANDLE(n) DT_PHANDLE_BY_IDX(DT_DRV_INST(n), partition, 0)
+
+#define DEFINE_FLASHDISKS_DEVICE(n)						\
+{										\
+	.info = {								\
+		.ops = &flash_disk_ops,						\
+		.name = DT_INST_PROP(n, disk_name),				\
+	},									\
+	.area_id = DT_FIXED_PARTITION_ID(PARTITION_PHANDLE(n)),			\
+	.offset = DT_REG_ADDR(PARTITION_PHANDLE(n)),				\
+	.size = DT_REG_SIZE(PARTITION_PHANDLE(n)),				\
+},
+
+static struct flashdisk_data flash_disks[] = {
+	DT_INST_FOREACH_STATUS_OKAY(DEFINE_FLASHDISKS_DEVICE)
 };
 
 static int disk_flash_init(const struct device *dev)
 {
 	ARG_UNUSED(dev);
+	int rc = 0;
+
+	for (int i = 0; i < ARRAY_SIZE(flash_disks); i++) {
+		rc |= disk_access_register(&flash_disks[i].info);
+	}
 
-	return disk_access_register(&flash_disk);
+	return rc;
```

Comment:
Shouldn't error be returned here only when all registrations fail?  The registration loop is called for all disks, whether previous registrations failed or not, so why fail completely when we had some success?